lib/auth/accountrecovery.go

```diff
@@ -373,6 +376,81 @@ func (s *Server) recordFailedRecoveryAttempt(ctx context.Context, username strin
 	return lockUntil, maxedAttempts, nil
 }
 
+// CompleteAccountRecovery implements AuthService.CompleteAccountRecovery.
+func (s *Server) CompleteAccountRecovery(ctx context.Context, req *proto.CompleteAccountRecoveryRequest) error {
+	if err := s.isAccountRecoveryAllowed(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	approvedToken, err := s.GetUserToken(ctx, req.GetRecoveryApprovedTokenID())
+	if err != nil {
+		log.Error(trace.DebugReport(err))
+		return trace.AccessDenied(completeRecoveryGenericErrMsg)
+	}
+
+	if err := s.verifyUserToken(approvedToken, UserTokenTypeRecoveryApproved); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check that the correct auth credential is being recovered before setting a new one.
+	switch req.GetNewAuthnCred().(type) {
+	case *proto.CompleteAccountRecoveryRequest_NewPassword:
+		if approvedToken.GetUsage() == types.UserTokenUsage_USER_TOKEN_RECOVER_MFA {
```

Comment:
Maybe:

```suggestion
		if approvedToken.GetUsage() != types.UserTokenUsage_USER_TOKEN_RECOVER_PASSWORD {
```

it's a bit safer on oneof changes, plus it matches the case ("case ... _NewPassword" and "if ... _PASSWORD"